Autounlock techwebs are shared between machines that use them.

JohnFulpWillard opened the pull request:
## About The Pull Request

Did you know that for every Autolathe, Limbgrower, Biogenerator, ORM, and Smelter that was built, an entire new techweb was made? The average round has 2 ORMs (smelters count) if not more from ORM deconstruction objective, 2 biogenerators, several autolathes and generally 0-1 limbgrower, I think there's more techwebs being created than we need.

Creating a whole techweb was a pretty terrible way to optimize this, so I made a global list that stores these techwebs. Created on demand, these autounlocking techwebs now share between the machines that use them. It also generate all hacked nodes which will be visible if the machine is hacked, instead of 'researching' them for each hacked individual machine.

The 'specialized' techweb subtype was removed because its sole purpose was to allow autounlocking to be a subtype of it. Now autounlocking is just the direct subtype. I also removed mechfab autounlocking type because it wasn't used, mechfabs use techwebs directly.

Autolathes and Limbgrowers now locally store 'imported designs' which are things uploaded from technology disks. Outside of this, the autounlocking techweb subtype now stores 'hacked designs' which unlocks when the machine is emagged.
While doing this, I saw ORMs and Biogenerators had disks you can insert into them, but I did not find anything that can actually be uploaded to them (I saw Alien Alloys as a possibility, but there's no such disk to allow uploading this alloy to the machine, and I didn't think an entire system for 1 single alloy was worth keeping around), so I removed this unused feature from both machines.

I merged 'Hacked' and 'Emagged' categories because they served the same purpose, 'hacked' being on hacked autolathes and 'emagged' being on hacked (emagged) limbgrowers.

Tech disk techwebs (which is created every time a disk is made, however I hope to change this in the future) into it's own subtype, so admins understand where its from when looking in VV panel (because I was confused when I saw them at first).

``autolathe_crafted`` proc was removed because it was entirely unused too.

Now it looks like this, which I consider an improvement:
![image](https://user-images.githubusercontent.com/53777086/211052875-9215a84d-fb04-450b-a120-4d0316cec6ff.png)

## Why It's Good For The Game

We no longer initialize a brand new techweb for no reason whatsoever.
Each techweb made is making entire lists of experiments and research papers, all of which is never to be seen in-game and is completely useless to the player. Cutting down on these techwebs is a good first step to this needless initializing.

## Changelog

:cl:
qol: Removed the tech disk part of the ORM's UI as it was entirely unused, now it's a little more compact with less scrolling needed.
refactor: Autolathes, Limb growers, Biogenerators, ORMs and Smelters now share techwebs with other machines of their types, rather than all make new techwebs each time. This means they only build their nodes once, including hacked ones. Instead of researching nodes on hacking the machine, they now show hacked ones depending on if it's hacked. 
/:cl:

JohnFulpWillard committed:
Autounlock techwebs are shared between machines that use them.

Did you know that for every Autolathe, Limbgrower, Biogenerator, ORM, and Smelter that was built, an entire new techweb was made?

Creating a whole techweb was a pretty terrible way to optimize this, so I made a global list that stores these techwebs. Created on demand, these autounlocking techwebs now share between the machines that use them.

I also removed 'specialized' because autounlocking was the only subtype of it and it had no unique featuers to it, as well as the exofab subtype, as they were completely unused (mechfabs just use their stored research node).

Lastly, I made tech disk techwebs (which is created every time a disk is made) into it's own subtype, so admins understand where its from when looking in VV panel (because I was confused when I saw them at first).

JohnFulpWillard commented on code/modules/research/techweb/all_nodes.dm:
```diff
@@ -2285,26 +2285,3 @@
-		TW.add_point_list(TN.research_costs)
-	return TW.research_points
-
-/proc/total_techweb_points_printout()
```
I was checking where techwebs are created and found these two procs which did nothing, so I just decided to remove them.

JohnFulpWillard committed:
DON'T DELETE THE TECHWEBS WHAT THE HELL

ZephyrTFA commented:
And how will this interact with machines being EMAGged. Won't this lead to emagging one emagging all similar types

JohnFulpWillard commented:
> And how will this interact with machines being EMAGged. Won't this lead to emagging one emagging all similar types

True, thanks for catching that

JohnFulpWillard committed:
Hacked and imported now works

Autolathes and Limbgrowers now locally store 'imported designs' which are things uploaded from technology disks. Outside of this, the autounlocking techweb subtype now stores 'hacked designs' which unlocks when the machine is emagged.

While doing this, I saw ORMs had disks you can insert into them, but I did not find anything that can actually be uploaded to them (I saw Alien Alloys as a possibility, but there's no such disk to allow uploading this alloy to the machine, and I didn't think an entire system for 1 single alloy was worth keeping around), so I removed it.

I also merged 'Hacked' and 'Emagged' categories because they served the same purpose, 'hacked' being on hacked autolathes and 'emagged' being on hacked (emagged) limbgrowers.

Lastly, I removed 'autolathe_crafted' proc because it was ALSO, entirely unused.

JohnFulpWillard committed:
Reworks importing & hacking

Hacked nodes are now added on add_design instead of snowflaking a new add_hacked_design

Importing now doesn't keep the same type of list as regular research techwebs because they aren't a techweb. Also re-adds the check to verify something is available to print before printing it.

finally, removes Biogenerator's design disk printing, because it ALSO was unused.

JohnFulpWillard committed:
Merge branch 'master' into autounlock

JohnFulpWillard committed:
don't path to the direct subpath

JohnFulpWillard committed:
fullupgrade isn't meant to be emagged

Done because I misread the original code.

ZephyrTFA commented on code/game/machinery/limbgrower.dm:
```diff
@@ -63,7 +75,14 @@
 		species_categories[species] = list()
-	for(var/design_id in stored_research.researched_designs)
+
+	var/list/available_nodes = stored_research.researched_designs
```
```suggestion
	var/list/available_nodes = stored_research.researched_designs.Copy()
```

JohnFulpWillard committed:
Update code/game/machinery/limbgrower.dm

Co-authored-by: Zephyr <12817816+ZephyrTFA@users.noreply.github.com>

ZephyrTFA reviewed (approved):
Looks good to me, currently cannot test it locally but I see no glaring issues

jlsnow301 commented on code/modules/research/techweb/techweb_types.dm:
```diff
@@ -0,0 +1,95 @@
+
+/datum/techweb/admin/New()
+	. = ..()
+	for(var/i in SSresearch.techweb_nodes)
+		var/datum/techweb_node/TN = SSresearch.techweb_nodes[i]
```
can you use `as anything` here

JohnFulpWillard replied:
We don't need ``as anything`` because we're not casting any type as ``i``

jlsnow301 replied:
sry i mean a for .. in loop rather, `for(var/datum/techweb_node/node in SSresearch.techweb_nodes)`

ZephyrTFA replied:
No we can't, because it's an assosciate list
